test/utils.dart

```diff
@@ -134,34 +134,52 @@ void _testFile(TestFile testFile) {
             indent: formatTest.leadingIndent,
             experimentFlags: const ['macros']);
 
-        var actual = formatter.formatSource(formatTest.input);
-
-        // The test files always put a newline at the end of the expectation.
-        // Statements from the formatter (correctly) don't have that, so add
-        // one to line up with the expected result.
-        var actualText = actual.text;
-        if (!testFile.isCompilationUnit) actualText += '\n';
-
-        // Fail with an explicit message because it's easier to read than
-        // the matcher output.
-        if (actualText != formatTest.output.text) {
-          fail('Formatting did not match expectation. Expected:\n'
-              '${formatTest.output.text}\nActual:\n$actualText');
-        } else if (actual.selectionStart != formatTest.output.selectionStart ||
-            actual.selectionLength != formatTest.output.selectionLength) {
-          fail('Selection did not match expectation. Expected:\n'
-              '${formatTest.output.textWithSelectionMarkers}\n'
-              'Actual:\n${actual.textWithSelectionMarkers}');
-        }
-
-        expect(actual.selectionStart, equals(formatTest.output.selectionStart));
-        expect(
-            actual.selectionLength, equals(formatTest.output.selectionLength));
+        var actual = _validateFormat(
+            formatter,
+            formatTest.input,
+            formatTest.output,
+            'did not match expectation',
+            testFile.isCompilationUnit);
+
+        // Make sure that formatting is idempotent. Format the output and make
+        // sure we get the same result.
+        _validateFormat(formatter, actual, formatTest.output,
```

Comment:
[optional] We know that `actual` and `formatTest.output` are equivalent given the above line passed, but my initial expectation reading the line performing a test for idempotency was to see `actual, actual` as the arguments.

```suggestion
        _validateFormat(formatter, actual, actual,
```

Reply:
Done, but I had to move some stuff around. The problem is that the actual doesn't *exactly* match the expectation when formatting a single statement because the actual output doesn't have a trailing newline (correct) while the expectation from the test file does (because each expectation ends on its own line).

Moved the code to handle this into TestFile so that the expectation is always exactly what we want.